include/pybind11/pytypes.h

```diff
@@ -1059,6 +1059,35 @@ inline str::str(const bytes& b) {
     m_ptr = obj.release().ptr();
 }
 
+/// \addtogroup pytypes
+/// @{
+class bytearray : public object {
+public:
+    PYBIND11_OBJECT(bytearray, object, PyByteArray_Check)
+
+    bytearray()
+        : object(PyByteArray_FromStringAndSize("", (ssize_t) 0), stolen_t{}) {
+        if (!m_ptr) pybind11_fail("Could not allocate bytearray object!");
+    }
+
+    bytearray(const char *c, size_t n)
+        : object(PyByteArray_FromStringAndSize(c, (ssize_t) n), stolen_t{}) {
+        if (!m_ptr) pybind11_fail("Could not allocate bytearray object!");
+    }
+
+    // Allow implicit conversion:
+    bytearray(const std::string &s) : bytearray(s.data(), s.size()) { }
+
+    operator std::string() const {
+        char *buffer = PyByteArray_AS_STRING(m_ptr);
```

Comment:
A NULL pointer would likely segfault. A pointer to, say, `PyLongObject`, reinterpreted as a pointer to `PyByteArrayObject` would be more dangerous, because that one might segfault, but also might just evaluate to corrupt data. I do not know how well should we try to protect the users from reckless casting...

Regarding the "macros don't check for NULL", despite what the documentation implies, the "regular" `PyByteArray_AsString()` also doesn't check anything. It has some `assert()`s, but those are turned off in regular use.

Reply:
> A NULL pointer would likely segfault.

Yes. The issue is that for some `py::object` subclasses, we allow `NULL` as valid, and for some not. And it's not very clear which is which, or what's the correct thing to do.

> A pointer to, say, `PyLongObject`, reinterpreted as a pointer to `PyByteArrayObject` would be more dangerous, because that one might segfault, but also might just evaluate to corrupt data. I do not know how well should we try to protect the users from reckless casting...

Yes, but that's why it's called `reinterpret_steal`/`reinterpret_borrow`? We've already fixed the constructors, but I do think that `reinterpret_*` shouts to the user _"you're reponsible"_. So I wouldn't worry about this; if you tell pybind11 that somethings a `bytearray`, who are we to question that?

> Regarding the "macros don't check for NULL", despite what the documentation implies, the "regular" `PyByteArray_AsString()` also doesn't check anything. It has some `assert()`s, but those are turned off in regular use.

Ugh; that's very ugly. But yes, now you say so, I do remember seeing that.